### PR TITLE
channeld: Verify the signature sent by the counterparty (aka do not trust verify)

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
This commit addresses an issue to enhance the resilience of core lightning when receiving node announcements.

According to BOLT 7 (The announcement_signatures Message), if the node_signature OR the bitcoin_signature is NOT correct, it is recommended to either send a warning and close the connection or send an error and fail the channel.

In this commit, we take a strict approach. If any error is detected, we send an error and fail the open channel operation. This is because the announcement_signatures operation is optional, and we assume that it must be correct.

lnprototest at commit dea47c29b5541dbfe7fe53cc2598330e897fa4f4 report the following error now.

```
2023-07-06T21:03:20.930Z DEBUG   hsmd: Shutting down

ERROR    root:helpers.py:170 Traceback (most recent call last):
  File "/home/vincent/Github/lightning/external/lnprototest/tests/helpers.py", line 167, in run_runner
    runner.run(test)
  File "/home/vincent/Github/lightning/external/lnprototest/lnprototest/runner.py", line 99, in run
    all_done = sequence.action(self)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/home/vincent/Github/lightning/external/lnprototest/lnprototest/structure.py", line 55, in action
    all_done &= e.action(runner)
                ^^^^^^^^^^^^^^^^
  File "/home/vincent/Github/lightning/external/lnprototest/lnprototest/event.py", line 365, in action
    raise EventError(self, "{}: message was {}".format(err, msg.to_str()))
lnprototest.errors.EventError: `Expected msgtype-warning, got msgtype-error: message was error channel_id=a37362839b13f61cfe82d35bd397b1264c389b245847cfb6111b38892546dc77 data=4661696c656420746f20766572696679206e6f64655f7369676e61747572652e` on event [{"event": "ExpectMsg", "file": "test_bolt2-01-close_channel.py", "pos": "157"},]
============================================================================================================================================================== short test summary info ===============================================================================================================================================================
FAILED tests/test_bolt2-01-close_channel.py::test_close_channel_shutdown_msg_normal_case_receiver_side - AssertionError: `Expected msgtype-shutdown, got msgtype-error: message was error channel_id=a37362839b13f61cfe82d35bd397b1264c389b245847cfb6111b38892546dc77 data=4661696c656420746f20766572696679206e6f64655f7369676e61747572652e` on event [{"event": "ExpectMsg", "file": "test_bolt2-01-close_channel.py", "pos": "75"},]
FAILED tests/test_bolt2-01-close_channel.py::test_close_channel_shutdown_msg_wrong_script_pubkey_receiver_side - AssertionError: `Expected msgtype-warning, got msgtype-error: message was error channel_id=a37362839b13f61cfe82d35bd397b1264c389b245847cfb6111b38892546dc77 data=4661696c656420746f20766572696679206e6f64655f7369676e61747572652e` on event [{"event": "ExpectMsg", "file": "test_bolt2-01-close_channel.py", "pos": "157"},]

```

This PR shows how useful is lnprototest in debugging issues and make sure that we are protocol compliant, so now lnprototest in this PR should fail, and when this is merged https://github.com/ElementsProject/lightning/pull/6383 the tests should be again happy

P.S: the current version of lnprototest is buggy because relay on cln behavior!

Changelog-Fixes: channeld: Verify the signature sent in announcement_signatures by the counterparty
Reported-by: lnprototest

### CODE REVIEW

**Comment by vincenzopalazzo** [comment]
Mh looks like this break our CI 

```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/test_pay.py::test_setchannel_restart - ValueError: Timeout while waiting for <function LightningNode.wait_channel_active.<locals>.<lambda> at 0x7f82a3753d40>
FAILED tests/test_invoices.py::test_invoice_routeboost_private - ValueError: Timeout while waiting for <function test_invoice_routeboost_private.<locals>.<lambda> at 0x7f82a1f86a70>
FAILED tests/test_plugin.py::test_sendpay_notifications - ValueError: Timeout while waiting for <function LightningNode.wait_channel_active.<locals>.<lambda> at 0x7f33910f75f0>
FAILED tests/test_plugin.py::test_coin_movement_notices - ValueError: Timeout while waiting for <function LightningNode.wait_channel_active.<locals>.<lambda> at 0x7f338fa009e0>
FAILED tests/test_opening.py::test_zeroconf_public - ValueError: Timeout while waiting for <function test_zeroconf_public.<locals>.<lambda> at 0x7f82a3aa2f80>
FAILED tests/test_pay.py::test_self_pay - ValueError: Timeout while waiting for <function NodeFactory.join_nodes.<locals>.<lambda> at 0x7f82a4d727a0>
FAILED tests/test_closing.py::test_onchain_close_upstream - ValueError: Timeout while waiting for <function LightningNode.wait_channel_active.<locals>.<lambda> at 0x7f338c3d5320>
FAILED tests/test_misc.py::test_listforwards_and_listhtlcs - ValueError: Timeout while waiting for <function LightningNode.wait_channel_active.<locals>.<lambda> at 0x7f82a1d70ef0>
FAILED tests/test_opening.py::test_zeroconf_multichan_forward - ValueError: Timeout while waiting for <function LightningNode.wait_channel_active.<locals>.<lambda> at 0x7f338f110ef0>
======= 9 failed, 51 passed, 2 skipped, 3 warnings in 1439.45s (0:23:59) =======
```

File: channeld/channeld.c [hunk 1]
```diff
@@ -669,6 +671,29 @@ static void handle_peer_announcement_signatures(struct peer *peer, const u8 *msg
...
 
+	/* BOLT 7:
+	 * - if the node_signature OR the bitcoin_signature is NOT correct:
+	 *  - MAY send a warning and close the connection, or send an error and fail the channel.
+	 *
+	 * In our case, we send an error and stop the open channel procedure. This approach is
+	 * considered overly strict since the peer can recover from it. However, this step is
+	 * optional. If the peer sends it, we assume that the signature must be correct.*/
+	assert(pubkey_from_node_id(&peer_pubkey, &peer->node_ids[REMOTE]));
+	sha256(&peer_announcement_signature, msg, tal_count(msg));
```
Comment (rustyrussell) [inline channeld/channeld.c:682]: This is wrong: the message it signs is not the announcement_signatures message, but the announcement message.

This requires a lot more logic!  You need to create the channel_announcement (with towire_channel_announcement, with 4 dummy signatures) and then replicate the logic in heck_channel_announcement() in gossipd/routing.c.

**Review by rustyrussell — CHANGES_REQUESTED** [review]

File: channeld/channeld.c (hunk 1 above)
Comment (vincenzopalazzo) [reply channeld/channeld.c:682]: Mh my bad I just assumed as a kid. I will check it now thanks!

**Comment by vincenzopalazzo** [comment]
Ok planning to finish it in the next 2 days

**Comment by vincenzopalazzo** [comment]
Ok this should be ready, I would like to wait for the CI to confirm

In addition, the following PR on lnprorotest shows that the current master fails https://github.com/rustyrussell/lnprototest/pull/100, but with this patch core lightning, it is 
able to pass the lnprototest tests again.

**Review by rustyrussell — APPROVED** [review]

**Comment by rustyrussell** [comment]
Ack, needs fixup folding...

**Comment by vincenzopalazzo** [comment]
Mh looks like this introduced a regression during a reconnect logic here are the logs https://gist.github.com/vincenzopalazzo/1a8cbb33e84d6b19af8e3e2a0b60a7af

Looks like that during the reconnect logic the short channel id changed?  and this can impact the announcement channels msg that will be signed? the full logs are available here https://github.com/ElementsProject/lightning/actions/runs/5561540390/jobs/10161624738?pr=6384

It should be the only regression introduced, any idea? @rustyrussell  

if my logic was wrong the test that requires to open a channel should not pass all?

**Comment by rustyrussell** [comment]
This broke dual funding, so clearly channel announcement isn't correct for that case?

**Review by rustyrussell — APPROVED** [review]
Ack 7f8193b56324c260661ee9c971f761e6b584e4dd

**Pull request merged**